### PR TITLE
BMP image loader

### PR DESCRIPTION
**Pull request by TurkeyMan**
Really simple and kinda lame BMP loader.

### CODE REVIEW

**Comment by drug007** [comment]
I'm pretty sure that existing simple implementation is better than non-existing full fledged one. So LGTM. But would be nice to have writing to file besides reading.

File: src/wg/format/bmp.d [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+        return inout(ImageBuffer)(
+            width, height,
+            (((bits * width) / 8) + 3) & ~3,
+            1, 1,
+            cast(ubyte)bits,
+            0,
+            data.ptr + header().offBits,
+            format.ptr,
+            null
+        );
```
Comment (igor84) [inline src/wg/format/bmp.d]: This is so unreadable without named parameters :(. Do we have some alternative to make this more readable?

Comment (p0nce) [reply src/wg/format/bmp.d]: if a parameter is unreadable, you can always have a named variable in the line before.

Comment (TurkeyMan) [reply src/wg/format/bmp.d]: Yeah D's `const` is a problem here; I couldn't just declare a mutable struct locally and populate it, and then return it... I had to initialise the const struct in-place.
I hate that code, I'll definitely rewrite most of it; but I just wanted something quick and dirty to get real pictures into the system for experiments.

**Comment by TurkeyMan** [comment]
For those reading; what I want to do here, is find a nice image loading API strategy where the image loader may be an 'ImageSource' like we've discussed before, that is, returns a range-based image-like object.
The logic there is that it won't eagerly load/decode the image into some temporary buffer that the user may just immediately convert into a different runtime form that they prefer. If the loader returns an ImageSource, then the user can chain their desired conversions and produce the loaded image into a buffer with their preferred formats and all that jazz directly.

I haven't shown anything like that here because we haven't talked about range-based image objects, and there's no design there yet; but what you can see here is some tech that will produce an appropriately formatted ImageBuffer pointing directly into the image memory, rather than reading/decoding the data into an allocated buffer. If it were RLE coded for instance, it would need to implement the range-like decode function when fetching scanlines or whatever.

**Comment by p0nce** [comment]
Mmmm, that would be quite elegant, also I can envision it will be way easier for BMP than JPEG!

For JPEG this would be probably hard, if there are entropy coding dependencies between 8x8 blocks? (some codec like HEIF will happily have a dependency over the whole image, such as you can't start unless at some "slice" points)

Presumably this sub-loading can be implemented over a lazily fully-decoded image buffer for complicated formats. Other than that it's clever as the Crop voldemort that would sub-rect the range will lead to less decoding, less allocation etc.

**Comment by rikkimax** [comment]
Truth be told, I think its trying to be a bit too smart so I'm not terribly keen on this idea.

**Comment by p0nce** [comment]
Other than that , I would strongly argue for doing _as few conversions as possible_ from the loaded data. In particular autoconverting to RGBA8 on load can be harmful.
(For example, in PDF embedding, RGBA8 and RGB8 PNG do embed differently, and it's necessary not to loose this information.) Auto-convert makes us less strong typed

**Comment by TurkeyMan** [comment]
> also I can envision it will be way easier for BMP than JPEG

Certainly; many formats will just demand a practical solution is to allocate a temp buffer, but there are lots of cases where that may be avoided.

**Comment by TurkeyMan** [comment]
@rikkimax In this BMP case, there's only 100 lines of BMP loading code here (mostly dealing with the fact there are 5 BMP header versions with different features), which gives you back the raw un-transformed image. It seems reasonably lean.

I could allocate a buffer and fill it, or do conversion by request on load, but that sort of could be performed by a range-like operation in this buffer in just one additional line.

**Comment by p0nce** [comment]
If the loader is to be a compile-time structure, then `loadPartial(rect)` could be an optional method.

**Comment by TurkeyMan** [comment]
A pipeline crop could do that...?

**Comment by TurkeyMan** [comment]
I added `void[] writeBMP(ref const(ImageBuffer) image)`
This is useful to experiment with this library in general, because it means you can write out images to something you can actually look at.

**Pull request merged**